File: Array/two-sum.py
# 60ms 75.80%
# 14.1MB 75.37%
from typing import List
import copy
class Solution:

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_tem = copy.deepcopy(nums)
        nums.sort()
        left = 0
        right = len(nums) - 1
        ans = []
        while left < right:

            if nums[left] + nums[right] > target:
                right -= 1
            elif nums[left] + nums[right] <target:
                left += 1
            else:
                for i in range(len(nums_tem)):
                    if nums[left] == nums_tem[i] or nums[right] == nums_tem[i]:
                        ans.append(i)
                break
        return ans

le = [3,2,4]
tar = 6
a = Solution()
b = a.twoSum(le, tar)
print(b)

File: Array/two-sum.cpp
#include <iostream>
#include <algorithm> // Include the <algorithm> header for lower_bound
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    struct Pair{
        int key;
        int value;
        bool operator < (const Pair &p) const{
            return value < p.value;
        }
        int operator + (const Pair &p) const{
            return value + p.value;
        }
    };

    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> ans;
        vector<Pair> p;
        for (int i = 0; i < nums.size(); i++){
            Pair temp;
            temp.key = i;
            temp.value = nums[i];
            p.push_back(temp);
        }
        sort(p.begin(), p.end());

        int left = 0, right = nums.size()-1;
        while (left < right){
            int sum = p[left] + p[right];
            if (sum < target){
                left += 1;
            } else if (sum > target){
                right -= 1;
            } else{
                ans = {p[left].key, p[right].key};
                break;
            }
        }
        return ans;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {3, 3};
    int target = 6;
    vector<int> ans = sol.twoSum(nums, target);
    for (int i = 0; i < ans.size(); i++){
        cout << ans[i] << " ";
    }
    cout << endl;
    return 0;
}